File: eBPF_Supermarket/CPU_Subsystem/BCC_sar/src/wakeup.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

#define PRINT_SLEEP 1
#define TARGET_PID 10327

BPF_ARRAY(symAddr, u64, 5);
BPF_STACK_TRACE(stacktraces, 4096);

struct sleep_record {
    u32 state;
    u32 in_sleep;
    u64 start;
};

BPF_HASH(sleepBegin, int, struct sleep_record, 4096);
BPF_HASH(sleeplast, int, u64, 4096);
BPF_HASH(runlast, int, u64, 4096);
BPF_HASH(runBegin, int, u64, 4096);
BPF_HASH(waitlast, int, u64, 4096);
BPF_HASH(waitBegin, int, u64, 4096);

BPF_PERF_OUTPUT(events);

struct wakeup_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    int prio;
    int success;
    int target_cpu;
};

struct wait_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    int prio;
};

struct sleep_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    u64 delay;
};

struct rq {
	raw_spinlock_t lock;
	unsigned int nr_running;
	unsigned int nr_numa_running;
	unsigned int nr_preferred_running;
	unsigned int numa_migrate_on;
	long unsigned int last_blocked_load_update_tick;
	unsigned int has_blocked_load;
};


struct cswch_args {
	unsigned long long pad;
	char prev_comm[16];
	pid_t prev_pid;
	int prev_prio;
	long prev_state;
	char next_comm[16];
	pid_t next_pid;
	int next_prio;
};

#define TASK_RUNNING			0x0000
#define TASK_INTERRUPTIBLE		0x0001
#define TASK_UNINTERRUPTIBLE		0x0002


#define BEGIN_SLEEP 1
#define END_SLEEP 2
#define BEGIN_WAIT 3
#define END_WAIT 4
#define BEGIN_RUN 5
#define END_RUN 6

struct perf_data {
    int stackid; // 调用栈的信息: waker/sleeper
    int pid;
    char comm[16];
    int type;
    int waker;
    char waker_comm[16];
    u64 time;
};


__always_inline static void copy_comm(char *src, char *dst) {
    for (int i = 0; i < 15; i++) {
        dst[i] = src[i];
    }
    dst[15] = 0;
}

// 事件统一用perf output形式传递到用户空间

// from trace_sched_switch
__always_inline static void begin_sleep(struct pt_regs *ctx) {
    struct task_struct *ts = bpf_get_current_task();
    u32 pid = ts->pid;
    struct perf_data data = {};

    data.stackid = stacktraces.get_stackid(ctx, 0);
    data.pid = pid;
    data.type = BEGIN_SLEEP;
    data.waker = 0;
    data.time = bpf_ktime_get_ns();

    // 获取进程的comm
    bpf_get_current_comm(&data.comm, sizeof(data.comm));

    // 向用户空间传送睡眠进程名、pid以及发生阻塞时的栈信息
    events.perf_submit(ctx, &data, sizeof(data));

    // bpf_trace_printk("#%-8d Begin Sleep", pid);
}

// #<pid>表示主语为pid
__always_inline static void end_sleep(struct pt_regs *waker_ctx, 
                                      struct wakeup_struct *ws, int waker_pid) {
    struct perf_data data = {};
    struct task_struct *ts = bpf_get_current_task();

    data.stackid = stacktraces.get_stackid(waker_ctx, 0);
    data.pid = ws->pid;
    data.type = END_SLEEP;
    data.waker = waker_pid;
    bpf_probe_read_kernel(data.waker_comm, 16, ts->comm);
    data.comm[15] = 0;
    copy_comm(ws->comm, data.comm);
    data.time = bpf_ktime_get_ns();

    events.perf_submit(waker_ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Wakeup %d", waker_pid, ws->pid);
    // bpf_trace_printk("#%-8d End Sleep", ws->pid);
}

__always_inline static void begin_run(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = BEGIN_RUN;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Begin Run", pid);
}

__always_inline static void end_run(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = END_RUN;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d End Run", pid);
}

__always_inline static void begin_wait(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = BEGIN_WAIT;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Begin Wait", pid);
}

__always_inline static void end_wait(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = END_WAIT;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d End Wait", pid);
}

int trace_sched_switch(struct cswch_args *ctx) {
    if (ctx->next_pid == ctx->prev_pid) return 0;

    struct task_struct *ts = bpf_get_current_task();
    struct sleep_record sr;
    u32 next = ctx->next_pid, prev = ctx->prev_pid;
    u64 time = bpf_ktime_get_ns(), *valp, delta, initval = 0;

    if ( (ts->state & TASK_INTERRUPTIBLE) || (ts->state & TASK_UNINTERRUPTIBLE) ) {
        // 进入睡眠状态
        sr.state = ts->state;
        sr.in_sleep = 1;
        sr.start = time;
        sleepBegin.update(&prev, &sr);

        begin_sleep((struct pt_regs *)ctx);
    } else if (ts->state == TASK_RUNNING) {
        // 开始等待
        waitBegin.update(&prev, &time);
        begin_wait(ctx->prev_pid, ctx->prev_comm, ctx);
    }

    // 更新新进程的开始执行时间
    runBegin.update(&next, &time);

    begin_run(ctx->next_pid, ctx->next_comm, ctx);
    end_run(ctx->prev_pid, ctx->prev_comm, ctx);

    // // next: 结束睡眠(一般不可能。一般要先被wakeup，不可能直接被调度运行)
    // struct sleep_record *p_sr;
    // p_sr = sleepBegin.lookup(&next);
    // if (p_sr && p_sr->in_sleep) {
    //     u64 last = time - p_sr->start;
    //     bpf_trace_printk("end sleep %d us\n", last);
    // }

    // next: 结束等待
    end_wait(ctx->next_pid, ctx->next_comm, ctx);
    valp = waitBegin.lookup(&next);
    if (valp && *valp != 0) {
        // bpf_trace_printk("valp = %llu\n", *valp);
        delta = time - *valp;

        valp = waitlast.lookup(&next);
        if (valp) *valp += delta;
        else waitlast.update(&next, &delta);

        waitBegin.update(&next, &initval);
    }

    // 统计旧进程的执行时间
    valp = runBegin.lookup(&prev);
    if (valp) {
        delta = time - *valp;
        valp = runlast.lookup(&prev);
        if (valp) *valp += delta;
        else runlast.update(&prev, &delta);
    }
    
    return 0;
}


int trace_wakeup(struct wakeup_struct *ws) {
    struct task_struct *ts = (struct task_struct *)bpf_get_current_task();
    // char comm[16];
    // bpf_probe_read_kernel(comm, 16, ts->comm);
    // comm[15] = 0;

    u32 pid = ws->pid;
    u64 *valp, nowtime = bpf_ktime_get_ns();
    struct sleep_record *p_sr = sleepBegin.lookup(&pid);
    if (p_sr && p_sr->in_sleep) {
        u64 time = nowtime - p_sr->start;
        // if (time > 1000000 && PRINT_SLEEP && pid == TARGET_PID) { // 1ms
        //     bpf_trace_printk("process %d sleep for %d us.(wakeup by %d)\n", pid, time / 1000, ts->pid);
        // }
        p_sr->in_sleep = 0;

        // 结束睡眠
        end_sleep((struct pt_regs *)ws, ws, ts->pid);

        // 开始等待
        waitBegin.update(&pid, &nowtime);
        begin_wait(pid, ws->comm, ws);

        valp = sleeplast.lookup(&pid);
        if (valp) *valp += time;
        else sleeplast.update(&pid, &time);
    }
    
    return 0;
}

// 在内核中，64位系统的 sizeof(long) == 8
int tick_update() {
    u32 key = 0;
    u64 *valp, v1 = 0, v2 = 0;
    valp = symAddr.lookup(&key);
    if(valp) v1 = *valp;

    key = 1;
    valp = symAddr.lookup(&key);
    if(valp) v2 = *valp;

    u64 percpu_offset[2];
    struct rq _rq;
    bpf_probe_read_kernel(percpu_offset, sizeof(u64) * 2, v1);
    u64 rq_ptr = v2 + percpu_offset[0]; // 默认读取CPU0的Rq
    bpf_probe_read_kernel(&_rq, sizeof(struct rq), rq_ptr);
    
    struct task_struct *ts = (struct task_struct *)bpf_get_current_task();
    int *pusage = &ts->usage; // usage表示进程被等待队列引用的数目
    // bpf_trace_printk("curr task usage = %d\n", *pusage);

    // 读取rq
    // bpf_trace_printk("%lx %lx\n", percpu_offset[0], percpu_offset[1]);
    // bpf_trace_printk("rqlen = %d\n", _rq.nr_running);
    return 0;
}


// 主要想评估一下是否wakeup成功，但是我觉得在trace_wake里面wake的程序应该都能成功
int kprobe_try_to_wake_up(struct pt_regs *ctx) {
    struct task_struct *p = PT_REGS_PARM1(ctx);
    struct task_struct *curr = bpf_get_current_task();
    bpf_trace_printk("try_to_wake_up %d, which state = %x", p->pid, p->state);

    return 0;
}

// 获取返回值
int kretprobe_try_to_wake_up(struct pt_regs *ctx) {
    // 从ctx里读取返回值
    int ret = PT_REGS_RC(ctx);

    bpf_trace_printk("[%d]try_to_wake_up", ret);
    return 0;
}

File: eBPF_Supermarket/CPU_Subsystem/BCC_sar/src/wakeup.py
#!/usr/bin/python3

from __future__ import print_function
from os import getpid
from time import sleep, time
from bcc import BPF, PerfType, PerfSWConfig
import bpfutil, ctypes, argparse

BEGIN_SLEEP = 1
END_SLEEP = 2
BEGIN_WAIT = 3
END_WAIT = 4
BEGIN_RUN = 5
END_RUN = 6

# 方便的管理数据随时间的更改，并能计算变化量
class deltaTimeMgr:
    def __init__(self, cnt: int) -> None:
        self.container = [0 for _ in range(cnt)]
        self.delta = [0 for _ in range(cnt)]
        self.cnt = cnt

    def update(self, lst) -> list:
        if len(lst) != self.cnt:
            raise Exception("列表长度不匹配")
        else:
            for i in range(len(lst)):
                self.delta[i] = lst[i] - self.container[i]
                self.container[i] = lst[i]
            return self.delta

def parse_args():
    parser = argparse.ArgumentParser(description="Wakeup Monitoring System", formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument("-t", "--type", default="time", help=
    """Type of output method.
event: print event to screen. (May cause lost of information)
time: print time of sleep,run,wait and compares their addition with actual time.
    -- Since our eBPF program is event driven, the static process like pid 1(systemd) 
which don't cause any event won't be recorded, so their value may become less.
    """)
    parser.add_argument("-p", "--pid", default=0, type=int, help="The Process needs to be attached on.")
    parser.add_argument("--time", default=3, type=float, help="excution time")
    parser.add_argument("-o", "--output", default="event.txt", type=str, help="The output file which saves all the content.")

    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()

    bpf = BPF(src_file="wakeup.c")
    bpf.attach_tracepoint(tp="sched:sched_wakeup", fn_name="trace_wakeup")
    bpf.attach_tracepoint(tp="sched:sched_switch", fn_name="trace_sched_switch")

    # 分别挂载在一个函数的进入和退出
    # bpf.attach_kprobe(event="try_to_wake_up", fn_name="kprobe_try_to_wake_up")
    # bpf.attach_kretprobe(event="try_to_wake_up", fn_name="kretprobe_try_to_wake_up")

    # 传递给bpf程序的定时事件
    bpf.attach_perf_event(ev_type=PerfType.SOFTWARE,
        ev_config=PerfSWConfig.CPU_CLOCK,
        fn_name="tick_update",
        sample_period=0, 
        sample_freq=3
    )

    # 内核中的重要变量定义，留作参考，勿删！
    # pcpu_base_addr = int(bpfutil.find_ksym("pcpu_base_addr"), base=16) # u32, Read Only
    # __per_cpu_start = int(bpfutil.find_ksym("__per_cpu_start"), base=16) # u32, Read Only
    # pcpu_unit_offsets = int(bpfutil.find_ksym("pcpu_unit_offsets"), base=16) # u32 array, Read Only

    # cpu偏移量
    __per_cpu_offset = int(bpfutil.find_ksym("__per_cpu_offset"), base=16) # u32 array, Inited
    # runqueues, 运行队列的percpu变量
    runqueues = int(bpfutil.find_ksym("runqueues"), base=16) # per_cpu var, u32

    bpf['symAddr'][0] = ctypes.c_ulonglong(__per_cpu_offset)
    bpf['symAddr'][1] = ctypes.c_ulonglong(runqueues)
    # 计算方式：某个cpu的PERCPU变量的地址 = PERCPU变量的地址 + __per_cpu_offset[cpu]

    # 传递自己的PID，使BPF避开这个pid，这件事非常重要！
    bpf["ownPid"][0] = ctypes.c_int(getpid())

    target_pid = args.pid
    nowtime = 0

    '''
    # 把栈信息写入到文件中
    def onexit():
        with open("sleep_stack.txt", "w") as f:
            for pid in pidSleepMap.keys():
                f.write("pid {}:\n".format(pid))
                for str in pidSleepMap[pid]:
                    f.write(str)
                    f.write("\n")
                f.write("\n")

        # 诀窍：读取schedule栈后(可用含有schedule来表示)的第一个函数
        with open("stack_type.txt", "w") as f:
            for stack in stack_set:
                f.write(stack)
                f.write("\n")

        print("Already Write.")
    atexit.register(onexit)
    '''

    if args.type == "time":
        thead = r"%runTime  %waitTime  %sleepTime  allTime/ms  elapsed/ms"
        print(thead)
        deltaMgr = deltaTimeMgr(4)
        while 1:
            runlast = waitlast = sleeplast = 0

            if ctypes.c_uint32(target_pid) in bpf["runlast"]: 
                runlast = bpf['runlast'][ctypes.c_uint32(target_pid)].value
            if ctypes.c_uint32(target_pid) in bpf["waitlast"]: 
                waitlast = bpf['waitlast'][ctypes.c_uint32(target_pid)].value
            if ctypes.c_uint32(target_pid) in bpf["sleeplast"]: 
                sleeplast = bpf['sleeplast'][ctypes.c_uint32(target_pid)].value
            all = runlast + waitlast + sleeplast

            runlast, waitlast, sleeplast, all = deltaMgr.update([runlast, waitlast, sleeplast, all])

            if all != 0:
                print("%8.2f  %9.2f  %9.2f  %10d  %10d" % 
                    (runlast / all, waitlast / all, sleeplast / all, int(all / 1000000), 1000) )
            else:
                print("%8.2f  %9.2f  %9.2f  %10d  %10d" % 
                    (0, 0, 0, int(all / 1000000), 1000) )
            
            sleep(1)
            nowtime += 1
            if nowtime % 12 == 0:
                print()
                print(thead)

    elif args.type == 'event':
        pidEvent = {} # 存储每个pid的event列表
        f = open(args.output, "w")
    
        # perf数据接收回调函数
        def print_event(cpu, data, size):
            global cnt
            cnt += 1
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            stackid = event.stackid

            comm = event.comm.decode('utf-8')
            waker_comm = event.waker_comm.decode('utf-8')

            # 读取栈信息到stack_str中
            if event.type <= 2:
                stack_str = ""
                # 直接打印整个栈的信息
                # for addr in bpf["stacktraces"].walk(stackid):
                #     sym = bpf.ksym(addr).decode('utf-8', 'replace')
                #     stack_str = stack_str + "\t" + sym + "\n"

                # 对于BEGIN_SLEEP, 打印wchan
                stack_str = ""
                if event.type == BEGIN_SLEEP:
                    find_schedule = 0
                    for addr in bpf["stacktraces"].walk(stackid):
                        sym = bpf.ksym(addr).decode('utf-8', 'replace')
                        if "schedule" in sym and find_schedule == 0:
                            find_schedule = 1
                        elif "schedule" not in sym and find_schedule == 1:
                            stack_str = sym
                            break
                else: # END_SLEEP
                    # 之后可能做一下数据分类，现在先只用第一项
                    for addr in bpf["stacktraces"].walk(stackid):
                        sym = bpf.ksym(addr).decode('utf-8', 'replace')
                        stack_str = sym
                        break
            else:
                stack_str = "None"

            if pid not in pidEvent:
                pidEvent[pid] = []

            pid_str = "#{}: {} {} type={} {} {} {}us\n{}\n\n".format(cnt, pid, comm, event.type, 
                        event.waker, waker_comm, int(event.time / 1000), stack_str)
            f.write(pid_str)
            pidEvent[pid].append(pid_str)

        # 准备从perf_output中接收数据
        start = time()
        bpf["events"].open_perf_buffer(print_event, page_cnt=128)
        global cnt
        cnt = 0

        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

            # nowtime += 1
            # print(nowtime)
            # if nowtime == 10:
            #     break

        f.close() # 关闭文件
        print("EVENT cnt = ", cnt)
        print("Tracepoint count = ", bpf["countMap"][0].value)

    elif args.type == "flame":
        f = open(args.output, "w")
        runbegin = {}
        def flame_record(cpu, data, size):
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            comm = event.comm.decode('utf-8')
            # waker_comm = event.waker_comm.decode('utf-8')

            if cpu != 0: return

            if event.type == BEGIN_RUN:
                runbegin[pid] = event.time
            elif event.type == END_RUN and pid in runbegin:
                if pid != 0:
                    time = int((event.time - runbegin[pid]) / 1000) # us
                else:
                    time = 3 # 优化idle进程的显示
                f.write("{0};{1} {2}\n".format(comm, pid, time))

        bpf["events"].open_perf_buffer(flame_record, page_cnt=128)
        start = time()
        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

    elif args.type == "runinterval":
        runbegin = {}
        def flame_record(cpu, data, size):
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            comm = event.comm.decode('utf-8')
            # waker_comm = event.waker_comm.decode('utf-8')

            if cpu != 0: return

            if event.type == BEGIN_RUN:
                runbegin[pid] = event.time
            elif event.type == END_RUN and pid in runbegin:
                if pid != 0:
                    time = int((event.time - runbegin[pid]) / 1000) # us
                else:
                    time = 3 # 优化idle进程的显示
                f.write("{0};{1} {2}\n".format(comm, pid, time))

        bpf["events"].open_perf_buffer(flame_record, page_cnt=128)
        start = time()
        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

    '''
    # perf缓冲区的用法(previous)
    def print_event(cpu, data, size):
        event = bpf["events"].event(data)
        print("recv msg.")

    # loop with callback to print_event
    bpf["events"].open_perf_buffer(print_event)

    while 1:
        bpf.perf_buffer_poll() # 仅适用于与poll不发生牵连影响的perf_submit
    '''